[SE-5180] UI for filtering OCIM instances

keithgg opened the pull request:
## Description

Adds UI filtering to the OCIM UI. Allows users to filter on the fields in keith/se_1142_filtering_on_ocim.

Developers are impacted.

## Supporting information

- JIRA: https://tasks.opencraft.com/browse/SE-1142.
![Screenshot 2022-01-18 at 08-09-42 Instances - OpenCraft -](https://user-images.githubusercontent.com/1470652/149880786-f3df072f-7e26-48bd-9249-e488a332ab96.png)
![Screenshot 2022-01-18 at 08-09-23 Instances - OpenCraft -](https://user-images.githubusercontent.com/1470652/149880791-022a6213-525d-4464-848a-59aba7b6682f.png)

## Testing instructions

- Log into the OCIM UI. 
- There will be a list of inputs that a user can filter on.
- In any of the text fields (Name, Notes) instances will be filtered on a partial match.
- For any of the dropdown fields (Status, OpenEdX Release) the instances will be filtered exactly.

## Deadline

## Other information

The UI for the Dropdowns isn't perfect. I didn't want to spend much more time on it, because
this UI will be rewritten in React at some point.

gabor-boros commented:
I've just seen this PR by accident... Let's merge it fast, I always wanted this feature :D

mtyaka commented on instance/api/appserver.py:
```diff
@@ -0,0 +1,57 @@
+
+class StatusViewSet(viewsets.ViewSet):
+    """
+    API to list all *unique* InstanceTag instances.
```
Appserver Statuses, not InstanceTags.

mtyaka commented on instance/api/deployment.py:
```diff
@@ -0,0 +1,57 @@
+
+class DeploymentTypeViewSet(viewsets.ViewSet):
+    """
+    API to list all *unique* InstanceTag instances.
```
DeploymentTypes, not InstanceTags.

mtyaka commented on instance/serializers/appserver.py:
```diff
@@ -67,3 +67,12 @@ def to_representation(self, instance):
+
+class StatusSerializer(serializers.Serializer):
+    """
+    Simple serializer for DeploymentType
```
Appserver Status, not DeploymentType.

mtyaka commented on instance/serializers/deployment.py:
```diff
@@ -0,0 +1,31 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Instance serializers (API representation)
```
DeploymentType, not Instance.

mtyaka commented on instance/tests/api/test_appserver.py:
```diff
@@ -0,0 +1,79 @@
+@patch(
+    'instance.tests.models.factories.openedx_instance.OpenEdXInstance._write_metadata_to_consul',
+    return_value=(1, True)
+)
```
We can probably remove these decorators since we're not using ddt or instance factories.

mtyaka commented on instance/tests/api/test_deployment.py:
```diff
@@ -0,0 +1,79 @@
+@patch(
+    'instance.tests.models.factories.openedx_instance.OpenEdXInstance._write_metadata_to_consul',
+    return_value=(1, True)
+)
```
We can probably remove these decorators since we're not using ddt or instance factories.

mtyaka commented on instance/tests/api/test_instance.py:
```diff
@@ -323,3 +324,53 @@ def test_set_notes_instance_gives_error_if_extra_field_is_provided(self, mock_co
+@patch(
+    'instance.tests.models.factories.openedx_instance.OpenEdXInstance._write_metadata_to_consul',
+    return_value=(1, True)
+)
```
We can probably remove these decorators since we're not using ddt or instance factories.

mtyaka commented on instance/tests/api/test_openedx_instance.py:
```diff
@@ -280,3 +284,229 @@ def test_get_details_unhealthy(self, mock_consul):
+@patch(
+    'instance.tests.models.factories.openedx_instance.OpenEdXInstance._write_metadata_to_consul',
+    return_value=(1, True)
+)
```
We can probably remove these decorators since we're not using ddt or instance factories.

keithgg replied:
Cool. Thought that was needed for something. Removed.

mtyaka commented on instance/api/appserver.py:
```diff
@@ -0,0 +1,57 @@
+        serializer = self.serializer_class(states, many=True)
+        return Response(serializer.data)
+
+    def retrieve(self, request, pk):
```
Do we need the endpoint to retrieve a single status? If not, we should probably just remove it.

keithgg replied:
Not necessarily. I put it in for completeness' sake.

mtyaka reviewed (commented):
Thanks for this @keithgg, the filters will make our lives much easier :)

I left a few minor comments, mostly about copy/pasted comments. I would like to test this on stage before approving it, but currently stage is in use.

This incorporates changes from https://github.com/open-craft/opencraft/pull/860 (which makes sense, since it depends on them), but I see it's using an older version of that code.  It might make sense to change this PR to target the [keith/se_1142_filtering_on_ocim](https://github.com/open-craft/opencraft/tree/keith/se_1142_filtering_on_ocim) branch instead of master.

keithgg commented:
Thanks for the review @mtyaka!

- I've updated the doc strings.
- Removed the retrieve views for Status and DeploymentType (you're right, they're not needed).
- I've merged in keith/se_1142_filtering_on_ocim

I don't expect to make any more changes on 1142, so if we keep this PR to master I think it's okay. If you disagree I'm cool with switching out the base branch.

mtyaka commented:
@keithgg I tested it on stage, it works really well! The production/sandbox filter doesn't work, but I'm guessing that's because the `PROD_APPSERVER_FAIL_EMAILS` variable is not set on stage.

The unit tests are currently failing, but once that's sorted out, this can be merged and deployed :)

:+1:

- [x] I tested this: I tested the filters on stage.
- [x] I read through the code
- [x] I checked for accessibility issues
- [ ] ~~Includes documentation~~

pomegranited commented:
@keithgg Ooo thank you for getting this feature done and merged! Very helpful UI, something we've wanted for a while!
